File: specs/service_bus_sas.py
"""Reference implementation of the SAS-over-REST send that specs/cope-service-bus-sas.feature automates.

The bulk-send-key primary key is read from the environment so it is never committed:
    setx SERVICE_BUS_SAS_KEY "<primary key from cope-pocsb > cope-requests > bulk-send-key>"
"""

import base64
import hashlib
import hmac
import json
import os
import time
import urllib.parse

import requests

NAMESPACE = os.environ.get("SERVICE_BUS_NAMESPACE", "cope-pocsb")
QUEUE_NAME = os.environ.get("COPE_QUEUE_NAME", "cope-requests")
POLICY_NAME = os.environ.get("SERVICE_BUS_SAS_KEY_NAME", "bulk-send-key")
POLICY_KEY = os.environ["SERVICE_BUS_SAS_KEY"]
CORRELATION_ID = os.environ.get("COPE_CORRELATION_ID", "caf-prospect-sas-rest")


def generate_sas_token(resource_uri: str, key_name: str, key: str, expiry_seconds: int = 3600) -> str:
    expiry = int(time.time() + expiry_seconds)
    encoded_uri = urllib.parse.quote_plus(resource_uri)
    string_to_sign = f"{encoded_uri}\n{expiry}".encode("utf-8")

    signature = base64.b64encode(
        hmac.new(key.encode("utf-8"), string_to_sign, hashlib.sha256).digest()
    )
    signature_encoded = urllib.parse.quote_plus(signature.decode("utf-8"))
    return f"SharedAccessSignature sr={encoded_uri}&sig={signature_encoded}&se={expiry}&skn={key_name}"



payload = {
    "requestMetadata": {
        "correlationId": CORRELATION_ID,
        "sourceSystem": "CaffeineCRM",
        "requestTimestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
    },
    "propertyAddress": {
        "line1": "4529 Winona Ct",
        "line2": "",
        "city": "Denver",
        "state": "CO",
        "zipCode": "80222",
    },
}

resource_uri = f"https://{NAMESPACE}.servicebus.windows.net/{QUEUE_NAME}"
sas_token = generate_sas_token(resource_uri, POLICY_NAME, POLICY_KEY)

response = requests.post(
    f"{resource_uri}/messages",
    headers={
        "Authorization": sas_token,
        "Content-Type": "application/json",
        # messageId / correlationId travel in BrokerProperties, not in the body.
        "BrokerProperties": json.dumps(
            {"MessageId": CORRELATION_ID, "CorrelationId": CORRELATION_ID}
        ),
    },
    params={"timeout": 60, "api-version": "2021-05"},
    data=json.dumps(payload),
    timeout=60,
)
print(response.status_code, response.text)

File: specs/support/service-bus-sas.js
const assert = require('node:assert/strict');
const crypto = require('node:crypto');
const Long = require('long');
const { ServiceBusClient } = require('@azure/service-bus');
const { createAzureCredential } = require('./azure-credential');

const DEFAULT_SAS_TTL_SECONDS = 3600;
const SERVICE_BUS_API_VERSION = '2021-05';

function namespaceHost(env = process.env, namespaceName) {
  const fqdn = env.SERVICE_BUS_NAMESPACE_FQDN?.trim();
  if (fqdn) return fqdn;

  assert.ok(
    namespaceName,
    'Set SERVICE_BUS_NAMESPACE_FQDN (e.g. cope-pocsb.servicebus.windows.net) to run SAS scenarios'
  );
  return `${namespaceName}.servicebus.windows.net`;
}

// Each queue carries its own "bulk-send-key" policy with a different key, so keys are looked up
// per queue first: SERVICE_BUS_SAS_KEY_COPE_REQUESTS_DEV for "cope-requests-dev", and so on.
function envSuffix(queueName) {
  return queueName.replace(/[^a-zA-Z0-9]+/g, '_').toUpperCase();
}

function resolveSasPolicy(env = process.env, policyName, queueName) {
  const suffix = queueName ? envSuffix(queueName) : null;
  const keyName =
    policyName ||
    (suffix && env[`SERVICE_BUS_SAS_KEY_NAME_${suffix}`]?.trim()) ||
    env.SERVICE_BUS_SAS_KEY_NAME?.trim();
  const key =
    (suffix && env[`SERVICE_BUS_SAS_KEY_${suffix}`]?.trim()) || env.SERVICE_BUS_SAS_KEY?.trim();

  assert.ok(keyName, 'Set SERVICE_BUS_SAS_KEY_NAME (e.g. bulk-send-key) to run SAS scenarios');
  assert.ok(
    key,
    `Set SERVICE_BUS_SAS_KEY_${suffix || '<QUEUE>'} to the primary key of the "${keyName}" policy on queue "${queueName}" to run SAS scenarios`
  );

  return { keyName, key };
}

// Mirrors the Service Bus SAS contract: HMAC-SHA256 over "<url-encoded-uri>\n<expiry>".
function generateSasToken(resourceUri, keyName, key, expirySeconds = DEFAULT_SAS_TTL_SECONDS) {
  const expiry = Math.floor(Date.now() / 1000) + expirySeconds;
  const encodedUri = encodeURIComponent(resourceUri);
  const signature = crypto
    .createHmac('sha256', key)
    .update(`${encodedUri}\n${expiry}`)
    .digest('base64');

  return {
    token: `SharedAccessSignature sr=${encodedUri}&sig=${encodeURIComponent(signature)}&se=${expiry}&skn=${encodeURIComponent(keyName)}`,
    expiry
  };
}

function buildQueueSasToken({ env = process.env, namespaceName, queueName, policyName, ttlSeconds } = {}) {
  const resourceUri = `https://${namespaceHost(env, namespaceName)}/${queueName}`;
  const { keyName, key } = resolveSasPolicy(env, policyName, queueName);
  const ttl = Number(ttlSeconds || env.SERVICE_BUS_SAS_TTL_SECONDS || DEFAULT_SAS_TTL_SECONDS);
  const { token, expiry } = generateSasToken(resourceUri, keyName, key, ttl);

  return { resourceUri, keyName, token, expiry };
}

function messagesUrl(resourceUri) {
  return `${resourceUri}/messages?timeout=60&api-version=${SERVICE_BUS_API_VERSION}`;
}

// The REST send API carries messageId/correlationId in the BrokerProperties header, not the body.
function brokerProperties({ messageId, correlationId, label }) {
  return JSON.stringify({
    ...(messageId ? { MessageId: messageId } : {}),
    ...(correlationId ? { CorrelationId: correlationId } : {}),
    ...(label ? { Label: label } : {})
  });
}

function createPeekClient(env = process.env, namespaceName) {
  const connection =
    env.SERVICE_BUS_LISTEN_CONNECTION?.trim() || env.SERVICE_BUS_CONNECTION?.trim();

  if (connection) {
    return { client: new ServiceBusClient(connection), authMode: 'connection string' };
  }

  return {
    client: new ServiceBusClient(namespaceHost(env, namespaceName), createAzureCredential(env)),
    authMode: 'Azure identity'
  };
}

// fromSequenceNumber is mandatory here: the SDK caches the last peeked sequence number per
// entity on the connection, so polling without it returns nothing after the first attempt.
async function peekFromStart(receiver, maxMessageCount) {
  try {
    return await receiver.peekMessages(maxMessageCount, { fromSequenceNumber: Long.ZERO });
  } finally {
    await receiver.close();
  }
}

async function peekQueue(client, queueName, maxMessageCount) {
  return peekFromStart(client.createReceiver(queueName, { receiveMode: 'peekLock' }), maxMessageCount);
}

async function peekDeadLetterQueue(client, queueName, maxMessageCount) {
  const receiver = client.createReceiver(queueName, {
    subQueueType: 'deadLetter',
    receiveMode: 'peekLock'
  });
  return peekFromStart(receiver, maxMessageCount);
}

function bodyAsText(message) {
  const body = message?.body;
  if (body === undefined || body === null) return '';
  if (typeof body === 'string') return body;
  if (Buffer.isBuffer(body)) return body.toString('utf8');
  return JSON.stringify(body);
}

module.exports = {
  SERVICE_BUS_API_VERSION,
  bodyAsText,
  brokerProperties,
  buildQueueSasToken,
  createPeekClient,
  generateSasToken,
  messagesUrl,
  namespaceHost,
  peekDeadLetterQueue,
  peekQueue
};
